pandas/core/arrays/interval.py

```diff
@@ -0,0 +1,856 @@
+            except TypeError:
+                msg = ('Cannot convert {dtype} to {new_dtype}; subtypes are '
+                       'incompatible')
+                raise TypeError(msg.format(dtype=self.dtype, new_dtype=dtype))
+            return self._shallow_copy(new_left, new_right)
+        elif is_interval_dtype(dtype):
+            if copy:
+                return self.copy()
+            else:
+                return self
+        elif is_categorical_dtype(dtype):
+            return Categorical(self.values)
+        # TODO: This try/except will be repeated.
+        try:
+            return self.values.astype(dtype, copy=copy)
+        except (TypeError, ValueError):
+            msg = 'Cannot cast {name} to dtype {dtype}'
+            raise TypeError(msg.format(name=type(self).__name__, dtype=dtype))
+
+    @classmethod
+    def _concat_same_type(cls, to_concat):
+        closed = set(interval.closed for interval in to_concat)
+        if len(closed) != 1:
+            raise ValueError("Intervals must all be closed on the same side.")
+        closed = closed.pop()
+
+        # TODO: avoid intermediate list
+        left = np.concatenate([interval.left for interval in to_concat])
+        right = np.concatenate([interval.right for interval in to_concat])
+        return cls._simple_new(left, right, closed=closed, copy=False)
+
+    def _shallow_copy(self, left=None, right=None):
+        from pandas.core.indexes.interval import IntervalIndex
+
+        if left is None:
+
+            # no values passed
+            # XXX: is ^ right? Or does that mean just left wasn't passed?
+            left, right = self.left, self.right
+
+        elif right is None:
+
+            # only single value passed, could be an IntervalIndex
+            # or array of Intervals
+            if not isinstance(left, (type(self), IntervalIndex)):
+                left = type(self)(left)
+
+            left, right = left.left, left.right
+        else:
+
+            # both left and right are values
+            pass
+
+        return self._simple_new(left, right, closed=self.closed,
+                                verify_integrity=False)
+
+    # TODO: doc
+    def copy(self, deep=False):
+        left = self.left.copy(deep=True) if deep else self.left
+        right = self.right.copy(deep=True) if deep else self.right
+        closed = self.closed
+        # TODO: Could skip verify_integrity here.
+        return type(self).from_arrays(left, right, closed=closed)
+
+    def _formatting_values(self):
+        return self.values
+
+    def get_values(self):
+        return self.values
+
+    def isna(self):
+        return isna(self.left)
+
+    @property
+    def nbytes(self):
+        return self.left.nbytes + self.right.nbytes
+
+    @property
+    def itemsize(self):
+        return self.left.itemsize + self.right.itemsize
+
+    def take(self, indices, axis=0, allow_fill=True, fill_value=None,
+             **kwargs):
```

Comment:
doc. wonder how of this you can push to EA base clasess (I suspect almost all of it), this is a pure indexing operation.

Reply:
I might be in favor of adding some basic implementation of `take` to EA, but in this case, `IntervalArray` would have to override it anyhow, as it needs to take both from `left` and `right`, which is something very specific to intervals.